plugins/mce/src/mceplugin.cpp

```diff
@@ -0,0 +1,191 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "common.h"
+#include "mceplugin.h"
+
+#include <voicecallmanagerinterface.h>
+
+#include <QtDBus>
+#include <QtPlugin>
+
+#define MCE_SERVICE "com.nokia.mce"
+#define MCE_PATH    "/com/nokia/mce/request"
+#define MCE_IFACE   "com.nokia.mce.request"
+
+class McePluginPrivate
+{
+    Q_DECLARE_PUBLIC(McePlugin)
+
+public:
+    McePluginPrivate(McePlugin *q)
+        : q_ptr(q),
+          manager(NULL)
+    { /* ... */ }
+
+    McePlugin *q_ptr;
+
+    VoiceCallManagerInterface *manager;
+
+    QHash<QString, AbstractVoiceCallHandler*> calls;
+};
+
+McePlugin::McePlugin(QObject *parent)
+    : AbstractVoiceCallManagerPlugin(parent),
+      d_ptr(new McePluginPrivate(this))
+{
+    TRACE
+}
+
+McePlugin::~McePlugin()
+{
+    TRACE
+    delete this->d_ptr;
+}
+
+QString McePlugin::pluginId() const
+{
+    TRACE
+    return PLUGIN_NAME;
+}
+
+bool McePlugin::initialize()
+{
+    TRACE
+    return true;
+}
+
+bool McePlugin::configure(VoiceCallManagerInterface *manager)
+{
+    TRACE
+    Q_D(McePlugin);
+
+    d->manager = manager;
+    return true;
+}
+
+bool McePlugin::start()
+{
+    TRACE
+    Q_D(const McePlugin);
+
+    QObject::connect(d->manager, SIGNAL(voiceCallsChanged()), SLOT(onVoiceCallsChanged()));
+    this->onVoiceCallsChanged();
+
+    return true;
+}
+
+bool McePlugin::suspend()
+{
+    TRACE
+    return true;
+}
+
+bool McePlugin::resume()
+{
+    TRACE
+    return true;
+}
+
+void McePlugin::finalize()
+{
+    TRACE
+}
+
+void McePlugin::onVoiceCallsChanged()
+{
+    TRACE
+    Q_D(McePlugin);
+
+    QString state = "none";
+    bool    isEmergency = false;
+
+    QDBusMessage message = QDBusMessage::createMethodCall(MCE_SERVICE, MCE_PATH, MCE_IFACE, "req_call_state_change");
+
+    if(d->manager->voiceCallCount() == 0)
+    {
+        d->calls.clear();
+    }
+    else
+    {
+        QStringList handlerIds;
+
+        // Go through call handlers and start processing status changes.
+        foreach(AbstractVoiceCallHandler *call, d->manager->voiceCalls())
+        {
+            handlerIds.append(call->handlerId());
+
+            if(!d->calls.contains(call->handlerId()))
+            {
+                DEBUG_T("Registering call handler from internal dictionary.");
+
+                QObject::connect(call, SIGNAL(statusChanged(VoiceCallStatus)), SLOT(onVoiceCallsChanged()));
+                d->calls.insert(call->handlerId(), call);
+            }
+
+            isEmergency |= call->isEmergency();
+
+            switch(call->status())
+            {
+            case AbstractVoiceCallHandler::STATUS_NULL:
+            case AbstractVoiceCallHandler::STATUS_DISCONNECTED:
+                continue;
+
+            case AbstractVoiceCallHandler::STATUS_INCOMING:
+                DEBUG_T("RINGING");
+                if(state != "active") state = "ringing";
```

Comment:
The other way around. In "active" it is perfectly normal for the display to be off (proximity / timeout / powerkey / whatnot). In "ringing" it will be turned on automatically if/once proximity sensor is not covered (and powekey / timeouts will not turn it off).